### PR TITLE
[Draft] Python: Promote the insecure cookie query from experimental

### PR DESCRIPTION
**Pull request by joefarebrother**
Part of https://github.com/github/codeql-python-team/issues/792 promoting https://github.com/github/codeql/pull/6360;
Depends on https://github.com/github/codeql/pull/16696

Promotes the insecure cookie query from experimental, finding instances of cookies set without secure values for the `Secure`, `HttpOnly`, or `SameSite` attributes.

### CODE REVIEW

File: python/ql/lib/semmle/python/Concepts.qll [hunk 1]
```diff
@@ -1231,6 +1246,104 @@ module Http {
...
+              b = true
+            )
+            or
+            exists(StringLiteral sl |
+              not sl.getText().regexpMatch("(?i).*;\\s*secure(;.*|\\s*)") and
+              DataFlow::localFlow(DataFlow::exprNode(sl), this.getHeaderArg()) and
+              b = false
+            )
+          )
+        }
```
Comment (yoff) [inline python/ql/lib/semmle/python/Concepts.qll:1268]:
```suggestion
        predicate hasSecureFlag(boolean b) {
          exists(StringLiteral sl |
            TaintTracking::localTaint(DataFlow::exprNode(sl), this.getHeaderArg())
          |
            if sl.getText().regexpMatch("(?i).*;\\s*secure(;.*|\\s*)") then b = true else b = false
          )
        }
```
I think this reads a bit easier: If there is in fact a literal defining the header argument, then we check it against the regexp and know the flag accordingly.
(The existence of `this.getHeaderArg()` is implied by the `localTaint` demand).

File: python/ql/lib/semmle/python/Concepts.qll [hunk 2]
```diff
@@ -1231,6 +1246,104 @@ module Http {
...
+              v instanceof SameSiteNone
+            )
+            or
+            exists(StringLiteral sl |
+              not sl.getText().regexpMatch("(?i).*;\\s*samesite=(strict|lax|none)(;.*|\\s*)") and
+              DataFlow::localFlow(DataFlow::exprNode(sl), this.getHeaderArg()) and
+              v instanceof SameSiteLax // Lax is the default
+            )
+          )
+        }
```
Comment (yoff) [inline python/ql/lib/semmle/python/Concepts.qll:1320]:
```suggestion
        predicate hasSameSiteAttribute(SameSiteValue v) {
          exists(StringLiteral sl |
            TaintTracking::localTaint(DataFlow::exprNode(sl), this.getHeaderArg())
          |
            sl.getText().regexpMatch("(?i).*;\\s*samesite=strict(;.*|\\s*)") and
            v instanceof SameSiteStrict
            or
            sl.getText().regexpMatch("(?i).*;\\s*samesite=lax(;.*|\\s*)") and
            v instanceof SameSiteLax
            or
            sl.getText().regexpMatch("(?i).*;\\s*samesite=none(;.*|\\s*)") and
            v instanceof SameSiteNone
            or
            not sl.getText().regexpMatch("(?i).*;\\s*samesite=(strict|lax|none)(;.*|\\s*)") and
            v instanceof SameSiteLax // Lax is the default
          )
        }
```
Similarly here: If there is a literal defining the header argument, we check it against a number of regexes and determine the value.

File: python/ql/lib/semmle/python/frameworks/Aiohttp.qll [hunk 3]
```diff
@@ -664,6 +664,51 @@ module AiohttpWebModel {
...
+          v instanceof Http::Server::CookieWrite::SameSiteLax
+          or
+          str.getText().toLowerCase() = "none" and
+          v instanceof Http::Server::CookieWrite::SameSiteNone
+        )
+      )
+      or
+      not exists(this.getArgByName("samesite")) and
+      v instanceof Http::Server::CookieWrite::SameSiteLax // Lax is the default
+    }
```
Comment (yoff) [inline python/ql/lib/semmle/python/frameworks/Aiohttp.qll:711]: This code seems to be copied across a bunch of frameworks. Would it make sense to create a `SetCookieCall` abstraction so the frameworks just have to point out where the calls are? It seems all calls to `set_cookie` (in any framework) are likely to follow this convention (and if not, it can be overridden by the framework).

**Review by yoff — CHANGES_REQUESTED** [review]
NIce work, I like the more careful semantics regarding known values.
A few suggestions around coding style.

File: python/ql/lib/semmle/python/Concepts.qll (hunk 1 above)
Comment (joefarebrother) [reply python/ql/lib/semmle/python/Concepts.qll:1268]: This does have a functional difference - currently, the `true` case uses `TaintTracking`, which allows for `"; Secure"` to be concatenated onto the cookie header, but the `false` case requires a `DataFlow` ensuring that we only report the flag as being not set when we know the exact string that is used. (in particular, we wouldn't report an alert in a case like `resp['Set-Cookie'] = "key=value; " + "Secure; " + "HttpOnly")`

File: python/ql/lib/semmle/python/frameworks/Aiohttp.qll (hunk 3 above)
Comment (joefarebrother) [reply python/ql/lib/semmle/python/frameworks/Aiohttp.qll:711]: Seems like it would make sense. Should it be defined in the `Concepts` library?

File: python/ql/lib/semmle/python/Concepts.qll (hunk 1 above)
Comment (yoff) [reply python/ql/lib/semmle/python/Concepts.qll:1268]: Ah! I overlooked that, that is nice, actually :-)

File: python/ql/lib/semmle/python/frameworks/Aiohttp.qll (hunk 3 above)
Comment (yoff) [reply python/ql/lib/semmle/python/frameworks/Aiohttp.qll:711]: Yes, since it comes from the protocol (something like [here]()https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie) it should probably not be in a specific framework, so if it could be in `Concepts`, that would be nice.

File: python/ql/lib/semmle/python/Concepts.qll (hunk 1 above)
Comment (yoff) [reply python/ql/lib/semmle/python/Concepts.qll:1268]: So, let me try again :-)
```suggestion
        predicate hasSecureFlag(boolean b) {
          exists(StringLiteral sl |
            // `sl` is likely a substring of the header
            TaintTracking::localTaint(DataFlow::exprNode(sl), this.getHeaderArg()) and
            sl.getText().regexpMatch("(?i).*;\\s*secure(;.*|\\s*)") and
            b = true
            or
            // `sl` is the entire header
            DataFlow::localFlow(DataFlow::exprNode(sl), this.getHeaderArg()) and
            not sl.getText().regexpMatch("(?i).*;\\s*secure(;.*|\\s*)") and
            b = false
          )
        }
```
So there is a literal in the code; either it gets to the header approximately and matches the regex, in which case the header now probably contains the string "; Secure", or it gets there exactly and does not match, in which case the header definitely does not contain the string "; Secure".

File: python/ql/lib/semmle/python/Concepts.qll (hunk 2 above)
Comment (yoff) [reply python/ql/lib/semmle/python/Concepts.qll:1320]: So with proper understanding of the semantics now, I might write it like this:
```suggestion
        predicate hasSameSiteAttribute(SameSiteValue v) {
          exists(StringLiteral sl |
            // `sl` is likely a substring of the header
            TaintTracking::localTaint(DataFlow::exprNode(sl), this.getHeaderArg()) and
            (
              sl.getText().regexpMatch("(?i).*;\\s*samesite=strict(;.*|\\s*)") and
              v instanceof SameSiteStrict
              or
              sl.getText().regexpMatch("(?i).*;\\s*samesite=lax(;.*|\\s*)") and
              v instanceof SameSiteLax
              or
              sl.getText().regexpMatch("(?i).*;\\s*samesite=none(;.*|\\s*)") and
              v instanceof SameSiteNone
            )
            or
            // `sl` is the entire header
            DataFlow::localFlow(DataFlow::exprNode(sl), this.getHeaderArg()) and
            not sl.getText().regexpMatch("(?i).*;\\s*samesite=(strict|lax|none)(;.*|\\s*)") and
            v instanceof SameSiteLax // Lax is the default
          )
        }
```

**Comment by yoff** [comment]
For the failing tests, it looks reasonable to just update the inline expectations. The content type will be set by the renderer in the unspecified cases, so is not statically known (at least not by our analysis).

File: python/ql/lib/semmle/python/Concepts.qll [hunk 4]
```diff
@@ -1203,6 +1203,75 @@ module Http {
...
+      /**
+       * Holds if the `SameSite` attribute of the cookie is known to have a value of `v`.
+       */
+      predicate hasSameSiteAttribute(CookieWrite::SameSiteValue v) { super.hasSameSiteAttribute(v) }
+    }
+
+    /**
+     * A dataflow call node to a method that sets a cookie in an http response,
+     * and has common keyword arguments `secure`, `httponly`, and `samesite` to set the attributes of the cookie.
+     */
```
Comment (yoff) [inline python/ql/lib/semmle/python/Concepts.qll:1226]:
```suggestion
    /**
     * A dataflow call node to a method that sets a cookie in an http response,
     * and has common keyword arguments `secure`, `httponly`, and `samesite` to set the attributes of the cookie.
     *
     * See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie
     */
```

**Review by yoff — APPROVED** [review]
Very nice! I have one suggestion for a doc link, but otherwise I think I cannot ask for more here :-)
We should kick off a DCA run, but I do not really expect any of this to blow up...

**Comment by yoff** [comment]
DCA run looks good. The new results look fine to me, but please confirm that you would expect them too :-) (The ones that are not unclassified; the unclassified ones are just our tests, it seems.)

**Review by yoff — APPROVED** [review]
LGTM
